### PR TITLE
refactor(data-types): convert data-types to TS

### PR DESCRIPTION
**Pull request by AllAwesome497**
### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [ ] Have you added new tests to prevent regressions?
- [ ] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [ ] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

converts data-types to TS.

I've typed this in a way that the types of columns can be extracted from `sequelize.define` (in theroy - haven't gotten far enough to test that yet).  I also added some generic types to assist in getting types of a DataType.

For any reviewers, I've labled 3 (ts) types which are relevant to each DataType:
- "Acceptable" type (ie what it'll accept as user input, e.g. for dates this is numbers, Dates, or strings)
- "Sane" type - What sequelize will normally return (e.g. for dates, Date)
- "Raw" type - Whatever we need to deserealize for a type.  (e.g. for a boolean this is a number, buffer, or string)

image

### Child PRs

- #13970
- #13981
- #14081
- #14220
- #14238
- #14239
- #14240

### The approach of this PR

Due to the size of this change, I've opted to split up this PR into many to make it more reviewable.

Here's the steps I'll be taking:
1. Create a "master" pr containing the changes which other data-types prs will be based on (this pr)
2. For each dialect, create a PR based on this branch containing the needed changes.
3. Once this PR is approved, merge in the child prs as they are approved.
4. Once all of the dialect specific code is in (meaning: all  lib/ code is approved and data-types should be working), start converting the tests to TS (in distinct prs, again for reviewability).
5. Once all test prs are approved & merged, merge this pr (assuming tests are green)

### CODE REVIEW

**Comment by WikiRik** [comment]
@AllAwesome497 checkout #12941 which was updated this week

**Comment by AllAwesome497** [comment]
> @AllAwesome497 checkout #12941 which was updated this week

Oh I completely missed that.

My change covers a few things that aren't handled in that pr such as named exports, inline static keys (e.g. `key` is then declared in the class itself rather than being exposed later), and more specific typing for types which vary from their raw values (in particular: "sane" vs "accepted" vs "raw" as I've called them).

**Comment by sdepold** [comment]
@jesse23 @Keimeno Up for a review?

**Comment by sdepold** [comment]
I think my TS skills don't really suffice to review this beast. I have an opinion the wording however:

"Acceptable" type (ie what it'll accept as user input, e.g. for dates this is numbers, Dates, or strings)
--> InputType

"Sane" type - What sequelize will normally return (e.g. for dates, Date)
--> SanitizedType

"Raw" type - Whatever we need to deserealize for a type. (e.g. for a boolean this is a number, buffer, or string)
--> OutputType

**Comment by sdepold** [comment]
Adding @ephys as well :)

**Comment by jesse23** [comment]
@AllAwesome497 2 questions in general...

- It is possible enable the test 1stly, then try to do pair converting with the test?  For example in this case, no need to convert all impacted test but at least the major `test/unit/sql/data-types.test.js`.

- Shall we merge with `data-types.d.ts` ( and eliminate `data-types.d.ts` eventually )?

**Review by ephys — COMMENTED** [review]
Looks good to me, this is my only note

File: lib/utils/class-to-invokable.ts
Comment (AllAwesome497) [reply lib/utils/class-to-invokable.ts:16]: This could also be `Array<any>` (and that's what I'd normally put).  I just did this since eslint recommended it and I was unsure if it'd work.  Will probably change it back to `Array<any>` since that's more intuitive IMO.

**Comment by AllAwesome497** [comment]
> I think my TS skills don't really suffice to review this beast. I have an opinion the wording however:
> 
> "Acceptable" type (ie what it'll accept as user input, e.g. for dates this is numbers, Dates, or strings) --> InputType
> 
> "Sane" type - What sequelize will normally return (e.g. for dates, Date) --> SanitizedType
> 
> "Raw" type - Whatever we need to deserealize for a type. (e.g. for a boolean this is a number, buffer, or string) --> OutputType

Sounds good, though I think that `RawType` is a bit clearer than `OutputType` since its more of an "internal" output type,

`Boolean` is one of the types where this is more relevant, since a "raw" boolean can be a buffer, string, or number imo.

**Comment by sdepold** [comment]
How are we doing on this front? I just wanted to point out one more time, that we are now targeting v7 and could hence drop support for any weird hack that was introduced earlier.

**Comment by AllAwesome497** [comment]
> How are we doing on this front? I just wanted to point out one more time, that we are now targeting v7 and could hence drop support for any weird hack that was introduced earlier.

Have been pretty busy recently, back on it for now.  

Looks like we'll need to re-model how `DataTypes` works but I don't think it'll be too bad (ie most standard usage should be fine iiuc).

**Comment by ephys** [comment]
I'm researching how we could support customising what type Sequelize must output for a given DataType. I'm waiting on this PR to be merged before properly experimenting but since you've knee-deep in DataTypes I'd like to know if my current draft would impact your decisions in this PR, like for `SaneTypeOf` https://gist.github.com/ephys/50a6a05be7322c64645be716fea6186a#support-for-alternative-js-types

Also I'd still like to bikeshed the names of the 3 types based on how I understood them:

- "Acceptable" -> UserInputTypeOf
- "Sane" SequelizeOutputTypeOf
- "Raw" -> SqlInputTypeOf

**Comment by AllAwesome497** [comment]
> I'm researching how we could support customising what type Sequelize must output for a given DataType. I'm waiting on this PR to be merged before properly experimenting but since you've knee-deep in DataTypes I'd like to know if my current draft would impact your decisions in this PR, like for `SaneTypeOf` https://gist.github.com/ephys/50a6a05be7322c64645be716fea6186a#support-for-alternative-js-types
> 
> Also I'd still like to bikeshed the names of the 3 types based on how I understood them:
> 
>     * "Acceptable" -> UserInputTypeOf
> 
>     * "Sane" SequelizeOutputTypeOf
> 
>     * "Raw" -> SqlInputTypeOf

Nope, I think your current draft is fine - though it will be a pain to migrate typing to this.

As I'm finding, there's a lot of undocumented dialect-specific logic around typing :/. 

As for renaming:  I will do this, I just haven't yet since there hasn't been a clear choice in terms of names yet (and I don't see this PR getting merged soon, given its scale so I figure its worth waiting).

**Comment by ephys** [comment]
> though it will be a pain to migrate typing to this.

Yup! Not looking forward to it but I want the feature enough to be motivated at least

**Comment by AllAwesome497** [comment]
> * 1stly

:thinking: I thought I responded to this but I don't see any response.

> It is possible enable the test 1stly, then try to do pair converting with the test? For example in this case, no need to convert all impacted test but at least the major test/unit/sql/data-types.test.js.

I've opted to convert everything to TS first and add tests later (however: this will be before this PR is merged).  I added some more details to the PR description.

> Shall we merge with data-types.d.ts ( and eliminate data-types.d.ts eventually )?

Yep!  That should be done by the time I merge this if I haven't already removed the file.

**Comment by AllAwesome497** [comment]
For renaming, thoughts on:

`AcceptedType` -> `InputType`
`SaneType` -> `OutputType`
`RawType` -> `ParsedType` 

CC @sdepold @ephys

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:503]: I think it would be a good opportunity to normalize `options`. We should either have `options`, or `_length` & co. But not both, otherwise we need to keep both in sync and it's error prone.

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:756]: use `isObject` from lodash instead

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:770]: same comment about deduplicating `options` & instance fields

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:667]: can't this be simplified to:

```suggestion
  public validate(value: AcceptedNumber): asserts value is AcceptedNumber {
```

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:783]: `AcceptedNumber` doesn't include string. But `DECIMAL` should accept strings to avoid loss of precision

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:799]: I know decorators are not standard yet but all this boilerplate would be much cleaner by using a `@newOptional` decorator.

This type of decorators is also possible with the updated decorator proposal so I think it's safe to use it internally, we'll just replace it when the time comes.

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:819]: value says it only accepts booleans (and falsy) but this tries to also accept things like `'true` and `'false'` and whatever object stringifies to these two strings.

I think we should restrict this one to only `true` and `false` and check using `typeof value === 'boolean'`

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:832]:
```suggestion
    if (value != null) {
```

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:903]: same comment about deduplicating `options` & instance fields

'_length' is probably used for the precision parameter most dialects support.

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:1036]: use `isPlainObject` from lodash instead

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:1120]: same comment about deduplicating

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:1130]: I don't think any dialect has `_` in their blob name https://new.sequelizejs.com/docs/v7/other-topics/other-data-types#blobs

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:1095]: BlobLength should be exported

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:1200]: we should avoid mutating options without making a copy first

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:1326]:
```suggestion
  returnType: ABSTRACT<T> | undefined;
  fields: string[] | undefined;
```

these fields are always created, but they can be undefiend

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:1427]: name note about deduplicating options

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:1420]: `DataTypes.ENUM('a', ['b', 'c'])` should throw. We should only support

```
DataTypes.ENUM('value', 'another value')
DataTypes.ENUM(['value', 'another value'])
DataTypes.ENUM({
  values: ['value', 'another value']
})
```

to avoid future accidental breaking changes because a refactor broke something someone accidentally used

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:1406]: `DataTypes.ENUM('a', ['b', 'c'])` shouldn't be accepted

```suggestion
  constructor(...members: Member[]);
```

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:1407]: this line doesn't need to be specified, unless you want TypeScript to allow users to use this overload. But I don't think they should.

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:1486]: same comment about deduplicating

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:1549]: same note about deduplicating

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:1545]: using a rest parameter here, when you don't actually accept rest params, is misleading. Plus this seems to be already covered by the previous 2 overloads

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:1826]: is this line necessary?

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:75]: doesn't `stripInternal` only apply to jsdoc? (either way this should be jsdoc)

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:23]: the purpose of these different types should be documented

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:19]: isn't this the same as the built-in type `InstanceType`?

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:15]:
```suggestion
const warnings: Record<string, boolean> = Object.create(null);
```

File: src/data-types.ts
Comment (ephys) [inline src/data-types.ts:2]: what is the reason for exposing these? Shouldn't users only use `DataTypes.x` and not `DataTypes.postgres.x`?

**Review by ephys — CHANGES_REQUESTED** [review]

**Comment by ephys** [comment]
Some comments may be unrelated to this PR

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:97]: Is there a reason to add it to the prototype too?

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:859]: I'd prefer to have the `_sanitize` instance method call the static `parse` method instead of getting something from the prototype

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:775]: if scale is specified but precision is not, we should throw with a clear error message

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:415]: Why were these changed, they should be `TINYTEXT` & co

File: src/dialects/abstract/data-types.ts
Comment (ephys) [inline src/dialects/abstract/data-types.ts:428]: We should remove these `return true` from validate.

**Review by ephys — CHANGES_REQUESTED** [review]

**Comment by AllAwesome497** [comment]
Closing in favor of https://github.com/sequelize/sequelize/pull/14505.

**Pull request closed without merging**
